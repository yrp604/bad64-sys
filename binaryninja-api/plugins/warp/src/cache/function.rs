use crate::convert::{comment_to_bn_comment, to_bn_symbol_at_address};
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::{Function as BNFunction, FunctionUpdateType};
use binaryninja::symbol::SymbolType;
use warp::signature::function::Function;

// TODO: Rename this?
/// Inserts a function match into the cache. This also has the side effect of setting persisted function
/// information, such as the matched function symbol.
///
/// IMPORTANT: This will mark the function as needing updates, if you intend to fill in functions with
/// no match (i.e. `None`), then you must change this function to prevent marking that as needing updates.
/// However, it's perfectly valid to remove a match and need to update the function still, so be careful.
pub fn insert_cached_function_match(function: &BNFunction, matched_function: Option<Function>) {
    let view = function.view();
    let function_start = function.start();
    // NOTE: If we expect to run match_function multiple times on a function, we should move this elsewhere.
    // Mark the function as needing updates so that reanalysis occurs on the function, and we apply the match.
    function.mark_updates_required(FunctionUpdateType::FullAutoFunctionUpdate);
    if let Some(auto_sym) = view.symbol_by_address(function_start) {
        // TODO: If we ever create non library function symbols we will need to remove this check (see: `to_bn_symbol_at_address`).
        if auto_sym.sym_type() == SymbolType::LibraryFunction {
            // NOTE: This will also mark for full auto function update, one thing to note is that the
            // requirement to call this is that this function not be called in the associated function's analysis.
            view.undefine_auto_symbol(&auto_sym);
        }
    }
    match matched_function {
        Some(matched_function) => {
            // Define the new matched function symbol, this can safely be done here as the symbol itself
            // will be persisted, unlike function type information or variable information.
            let new_sym = to_bn_symbol_at_address(&view, &matched_function.symbol, function_start);
            view.define_auto_symbol(&new_sym);
            // TODO: How to clear the comments? They are just persisted.
            // TODO: Also they generate an undo action, i hate implicit undo actions so much.
            for comment in &matched_function.comments {
                let bn_comment = comment_to_bn_comment(&function, comment.clone());
                function.set_comment_at(bn_comment.addr, &bn_comment.comment);
            }
            function.store_metadata("warp_matched_function", &matched_function.to_bytes(), false);
        }
        None => {
            function.remove_metadata("warp_matched_function");
        }
    }
}

// TODO: This does allocations, and for every reanalysis.
pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let metadata = function.query_metadata("warp_matched_function")?;
    let raw_metadata = metadata.get_raw()?;
    Function::from_bytes(&raw_metadata)
}
