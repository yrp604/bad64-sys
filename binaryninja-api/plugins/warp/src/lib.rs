use crate::cache::{
    cached_adjacency_constraints, cached_call_site_constraints, cached_function_guid,
};
use crate::convert::{from_bn_symbol, from_bn_type};
use binaryninja::architecture::{
    Architecture, ImplicitRegisterExtend, Register as BNRegister, RegisterInfo,
};
use binaryninja::basic_block::BasicBlock as BNBasicBlock;
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::confidence::MAX_CONFIDENCE;
use binaryninja::function::{Function as BNFunction, NativeBlock};
use binaryninja::low_level_il::expression::{ExpressionHandler, LowLevelILExpressionKind};
use binaryninja::low_level_il::function::{
    FunctionMutability, LowLevelILFunction, NonSSA, RegularNonSSA,
};
use binaryninja::low_level_il::instruction::{
    InstructionHandler, LowLevelILInstruction, LowLevelILInstructionKind,
};
use binaryninja::low_level_il::{LowLevelILRegister, VisitorAction};
use binaryninja::rc::Ref as BNRef;
use std::path::PathBuf;
use warp::signature::basic_block::BasicBlockGUID;
use warp::signature::function::constraints::FunctionConstraints;
use warp::signature::function::{Function, FunctionGUID};

pub mod cache;
pub mod convert;
mod matcher;
/// Only used when compiled for cdylib target.
mod plugin;

pub fn core_signature_dir() -> PathBuf {
    // Get core signatures for the given platform
    let install_dir = binaryninja::install_directory();
    // macOS core dir is separate from the install dir.
    #[cfg(target_os = "macos")]
    let core_dir = install_dir.parent().unwrap().join("Resources");
    #[cfg(not(target_os = "macos"))]
    let core_dir = install_dir;
    core_dir.join("signatures")
}

pub fn user_signature_dir() -> PathBuf {
    binaryninja::user_directory().join("signatures/")
}

pub fn build_function<A: Architecture, M: FunctionMutability>(
    func: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> Function {
    let bn_fn_ty = func.function_type();
    Function {
        guid: cached_function_guid(func, llil),
        symbol: from_bn_symbol(&func.symbol()),
        ty: from_bn_type(&func.view(), &bn_fn_ty, MAX_CONFIDENCE),
        constraints: FunctionConstraints {
            // NOTE: Adding adjacent only works if analysis is complete.
            // NOTE: We do not filter out adjacent functions here.
            adjacent: cached_adjacency_constraints(func, |_| true),
            call_sites: cached_call_site_constraints(func),
            // TODO: Add caller sites (when adjacent and call sites are minimal)
            // NOTE: Adding caller sites only works if analysis is complete.
            caller_sites: Default::default(),
        },
    }
}

/// Basic blocks sorted from high to low.
pub fn sorted_basic_blocks(func: &BNFunction) -> Vec<BNRef<BNBasicBlock<NativeBlock>>> {
    let mut basic_blocks = func
        .basic_blocks()
        .iter()
        .map(|bb| bb.clone())
        .collect::<Vec<_>>();
    basic_blocks.sort_by_key(|f| f.start_index());
    basic_blocks
}

pub fn function_guid<A: Architecture, M: FunctionMutability>(
    func: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> FunctionGUID {
    let basic_blocks = sorted_basic_blocks(func);
    let basic_block_guids = basic_blocks
        .iter()
        .map(|bb| basic_block_guid(bb, llil))
        .collect::<Vec<_>>();
    FunctionGUID::from_basic_blocks(&basic_block_guids)
}

pub fn basic_block_guid<A: Architecture, M: FunctionMutability>(
    basic_block: &BNBasicBlock<NativeBlock>,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> BasicBlockGUID {
    let func = basic_block.function();
    let view = func.view();
    let arch = func.arch();
    let max_instr_len = arch.max_instr_len();

    // NOPs and useless moves are blacklisted to allow for hot-patchable functions.
    let is_blacklisted_instr = |instr: &LowLevelILInstruction<A, M, NonSSA<RegularNonSSA>>| {
        match instr.kind() {
            LowLevelILInstructionKind::Nop(_) => true,
            LowLevelILInstructionKind::SetReg(op) => {
                match op.source_expr().kind() {
                    LowLevelILExpressionKind::Reg(source_op)
                        if op.dest_reg() == source_op.source_reg() =>
                    {
                        match op.dest_reg() {
                            LowLevelILRegister::ArchReg(r) => {
                                // If this register has no implicit extend then we can safely assume it's a NOP.
                                // Ex. on x86_64 we don't want to remove `mov edi, edi` as it will zero the upper 32 bits.
                                // Ex. on x86 we do want to remove `mov edi, edi` as it will not have a side effect like above.
                                matches!(
                                    r.info().implicit_extend(),
                                    ImplicitRegisterExtend::NoExtend
                                )
                            }
                            LowLevelILRegister::Temp(_) => false,
                        }
                    }
                    _ => false,
                }
            }
            _ => false,
        }
    };

    let is_variant_instr = |instr: &LowLevelILInstruction<A, M, NonSSA<RegularNonSSA>>| {
        let is_variant_expr = |expr: &LowLevelILExpressionKind<A, M, NonSSA<RegularNonSSA>>| {
            // TODO: Checking the section here is slow, we should gather all section ranges outside of this.
            match expr {
                LowLevelILExpressionKind::ConstPtr(op)
                    if !view.sections_at(op.value()).is_empty() =>
                {
                    // Constant Pointer must be in a section for it to be relocatable.
                    // NOTE: We cannot utilize segments here as there will be a zero based segment.
                    true
                }
                LowLevelILExpressionKind::ExternPtr(_) => true,
                LowLevelILExpressionKind::Const(op) if !view.sections_at(op.value()).is_empty() => {
                    // Constant value must be in a section for it to be relocatable.
                    // NOTE: We cannot utilize segments here as there will be a zero based segment.
                    true
                }
                _ => false,
            }
        };

        // Visit instruction expressions looking for variant expression, [VisitorAction::Halt] means variant.
        instr.visit_tree(&mut |expr| {
            if is_variant_expr(&expr.kind()) {
                // Found a variant expression
                VisitorAction::Halt
            } else {
                VisitorAction::Descend
            }
        }) == VisitorAction::Halt
    };

    let basic_block_range = basic_block.start_index()..basic_block.end_index();
    let mut basic_block_bytes = Vec::with_capacity(basic_block_range.count());
    for instr_addr in basic_block.into_iter() {
        let mut instr_bytes = view.read_vec(instr_addr, max_instr_len);
        if let Some(instr_info) = arch.instruction_info(&instr_bytes, instr_addr) {
            instr_bytes.truncate(instr_info.length);
            if let Some(instr_llil) = llil.instruction_at(instr_addr) {
                // If instruction is blacklisted don't include the bytes.
                if !is_blacklisted_instr(&instr_llil) {
                    if is_variant_instr(&instr_llil) {
                        // Found a variant instruction, mask off entire instruction.
                        instr_bytes.fill(0);
                    }
                    // Add the instructions bytes to the basic blocks bytes
                    basic_block_bytes.extend(instr_bytes);
                }
            }
        }
    }

    BasicBlockGUID::from(basic_block_bytes.as_slice())
}

#[cfg(test)]
mod tests {
    use crate::cache::cached_function_guid;
    use binaryninja::binary_view::BinaryViewExt;
    use binaryninja::headless::Session;
    use std::path::PathBuf;
    use std::sync::OnceLock;

    static INIT: OnceLock<Session> = OnceLock::new();

    fn get_session<'a>() -> &'a Session {
        // TODO: This is not shared between other test modules, should still be fine (mutex in core now).
        INIT.get_or_init(|| Session::new().expect("Failed to initialize session"))
    }

    #[test]
    fn insta_signatures() {
        let session = get_session();
        let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
        for entry in std::fs::read_dir(out_dir).expect("Failed to read OUT_DIR") {
            let entry = entry.expect("Failed to read directory entry");
            let path = entry.path();
            if path.is_file() {
                let view = session.load(&path).expect("Failed to load view");
                let mut functions = view
                    .functions()
                    .iter()
                    .map(|f| cached_function_guid(&f, &f.low_level_il().unwrap()))
                    .collect::<Vec<_>>();
                functions.sort_by_key(|guid| guid.guid);
                let snapshot_name =
                    format!("snapshot_{}", path.file_stem().unwrap().to_string_lossy());
                insta::assert_debug_snapshot!(snapshot_name, functions);
            }
        }
    }
}
