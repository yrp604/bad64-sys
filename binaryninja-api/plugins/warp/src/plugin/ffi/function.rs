use crate::build_function;
use crate::cache::{insert_cached_function_match, try_cached_function_match};
use crate::convert::{to_bn_symbol_at_address, to_bn_type};
use crate::plugin::ffi::{BNWARPConstraint, BNWARPFunction, BNWARPFunctionGUID};
use binaryninja::function::Function;
use binaryninja::rc::Ref;
use binaryninja::string::BnString;
use binaryninjacore_sys::{BNFunction, BNSymbol, BNType};
use std::ffi::c_char;
use std::mem::ManuallyDrop;
use std::sync::Arc;
use warp::signature::comment::FunctionComment;

#[repr(C)]
pub struct BNWarpFunctionComment {
    pub text: *mut c_char,
    pub offset: i64,
}

impl BNWarpFunctionComment {
    /// Leaks the text string to be freed with BNWARPFreeFunctionComment
    pub fn from_owned(value: &FunctionComment) -> Self {
        let text = BnString::into_raw(BnString::new(&value.text));
        Self {
            text,
            offset: value.offset,
        }
    }

    pub unsafe fn free_raw(value: &Self) {
        unsafe { BnString::free_raw(value.text) }
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match build_function(&function, || function.lifted_il().ok(), false) {
        Some(function) => Arc::into_raw(Arc::new(function)) as *mut BNWARPFunction,
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetMatchedFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match try_cached_function_match(&function) {
        Some(matched_function) => {
            let arc_matched_function = Arc::new(matched_function);
            // NOTE: Freed by BNWARPFreeFunctionReference
            Arc::into_raw(arc_matched_function) as *mut BNWARPFunction
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionApply(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) {
    let analysis_function = Function::from_raw(analysis_function);
    match function.is_null() {
        false => {
            // Set the matched function to `function`.
            let matched_function = ManuallyDrop::new(Arc::from_raw(function));
            insert_cached_function_match(
                &analysis_function,
                Some(matched_function.as_ref().clone()),
            )
        }
        true => {
            // We are removing the previous match.
            insert_cached_function_match(&analysis_function, None)
        }
    };
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetGUID(
    function: *mut BNWARPFunction,
) -> BNWARPFunctionGUID {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    function.guid
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbol(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNSymbol {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let view = analysis_function.view();
    let address = analysis_function.symbol().address();
    let function_symbol = to_bn_symbol_at_address(&view, &function.symbol, address);
    // NOTE: The symbol ref has been pre-incremented for the caller.
    Ref::into_raw(function_symbol).handle
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbolName(function: *mut BNWARPFunction) -> *mut c_char {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let bn_name = BnString::new(&function.symbol.name);
    // NOTE: The symbol name string to be freed by BNFreeString
    BnString::into_raw(bn_name)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetType(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNType {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    match &function.ty {
        Some(func_ty) => {
            let arch = analysis_function.arch();
            let function_type = to_bn_type(Some(arch), func_ty);
            // NOTE: The type ref has been pre-incremented for the caller.
            unsafe { Ref::into_raw(function_type) }.handle
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetConstraints(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWARPConstraint {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_constraints: Box<[BNWARPConstraint]> = function
        .constraints
        .clone()
        .into_iter()
        .map(Into::into)
        .collect();
    *count = raw_constraints.len();
    let raw_constraints_ptr = Box::into_raw(raw_constraints);
    raw_constraints_ptr as *mut BNWARPConstraint
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetComments(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWarpFunctionComment {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_comments: Box<[_]> = function
        .comments
        .iter()
        .map(BNWarpFunctionComment::from_owned)
        .collect();
    *count = raw_comments.len();
    let raw_comments_ptr = Box::into_raw(raw_comments);
    raw_comments_ptr as *mut BNWarpFunctionComment
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionsEqual(
    function_a: *mut BNWARPFunction,
    function_b: *mut BNWARPFunction,
) -> bool {
    // We do not own function so we should not drop.
    let function_a = ManuallyDrop::new(Arc::from_raw(function_a));
    // We do not own function so we should not drop.
    let function_b = ManuallyDrop::new(Arc::from_raw(function_b));
    function_a.eq(&function_b)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPNewFunctionReference(
    function: *mut BNWARPFunction,
) -> *mut BNWARPFunction {
    Arc::increment_strong_count(function);
    function
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionReference(function: *mut BNWARPFunction) {
    if function.is_null() {
        return;
    }
    Arc::decrement_strong_count(function);
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionList(functions: *mut *mut BNWARPFunction, count: usize) {
    let functions_ptr = std::ptr::slice_from_raw_parts_mut(functions, count);
    let functions = Box::from_raw(functions_ptr);
    for function in functions {
        // NOTE: The functions themselves should also be arc.
        BNWARPFreeFunctionReference(function);
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionCommentList(
    comments: *mut BNWarpFunctionComment,
    count: usize,
) {
    let comments_ptr = std::ptr::slice_from_raw_parts_mut(comments, count);
    let comments = Box::from_raw(comments_ptr);
    for comment in &comments {
        BNWarpFunctionComment::free_raw(comment)
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeConstraintList(
    constraints: *mut BNWARPConstraint,
    count: usize,
) {
    let constraints_ptr = std::ptr::slice_from_raw_parts_mut(constraints, count);
    let _constraints = unsafe { Box::from_raw(constraints_ptr) };
}
