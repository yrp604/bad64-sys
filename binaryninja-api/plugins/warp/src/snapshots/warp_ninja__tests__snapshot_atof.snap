---
source: plugins/warp/src/lib.rs
expression: functions
---
[
    FunctionGUID {
        guid: 02d864ef-80bc-5265-b78f-0e42815a4d1d,
    },
    FunctionGUID {
        guid: 02d864ef-80bc-5265-b78f-0e42815a4d1d,
    },
    FunctionGUID {
        guid: 06a9339a-7249-50f3-9c3d-131d3248d560,
    },
    FunctionGUID {
        guid: 06a9339a-7249-50f3-9c3d-131d3248d560,
    },
    FunctionGUID {
        guid: 06a9339a-7249-50f3-9c3d-131d3248d560,
    },
    FunctionGUID {
        guid: 06a9339a-7249-50f3-9c3d-131d3248d560,
    },
    FunctionGUID {
        guid: 07c4bac9-cf7d-5e00-833d-d23ba44c07d9,
    },
    FunctionGUID {
        guid: 0916139e-a5db-51c8-be24-2c6fb20a9e3a,
    },
    FunctionGUID {
        guid: 0c7b078d-dafb-5be7-8b10-a25b1e691b33,
    },
    FunctionGUID {
        guid: 0e6feb85-ff6f-5b86-bd2c-26fb9829241c,
    },
    FunctionGUID {
        guid: 0e6feb85-ff6f-5b86-bd2c-26fb9829241c,
    },
    FunctionGUID {
        guid: 11592e15-a2a6-51b3-bda2-534cc0434df0,
    },
    FunctionGUID {
        guid: 13b16f81-0c6f-5aee-ad9f-0d142658cf18,
    },
    FunctionGUID {
        guid: 1b1669c2-4a81-5300-99f7-14cba1cc7b24,
    },
    FunctionGUID {
        guid: 1b969932-96ce-5a6f-b91c-82e5f7b7f33d,
    },
    FunctionGUID {
        guid: 1c037c1e-9af1-59af-b19e-122fdf885148,
    },
    FunctionGUID {
        guid: 1de88f85-a19c-5454-9122-ddc80f7509b4,
    },
    FunctionGUID {
        guid: 1de88f85-a19c-5454-9122-ddc80f7509b4,
    },
    FunctionGUID {
        guid: 1e794537-6289-59e7-bef9-0c72f3989db8,
    },
    FunctionGUID {
        guid: 1eee5e05-95ef-5f99-8713-504c2003183c,
    },
    FunctionGUID {
        guid: 2833f544-9acb-552d-aa4b-df7d51dbab64,
    },
    FunctionGUID {
        guid: 29690354-fa27-54d1-a8be-18535b19b1c3,
    },
    FunctionGUID {
        guid: 31a8e73e-74fe-5a33-b0ab-aa1b136e021d,
    },
    FunctionGUID {
        guid: 364c5864-0cd1-5527-8ca0-2378d5b4c0a5,
    },
    FunctionGUID {
        guid: 3c38a678-9c4e-527a-b543-3075b345d532,
    },
    FunctionGUID {
        guid: 3c38a678-9c4e-527a-b543-3075b345d532,
    },
    FunctionGUID {
        guid: 3cc2e827-b707-5c9d-82e9-d76ac3abc904,
    },
    FunctionGUID {
        guid: 3eb2a4c5-84f5-5687-92a0-35ef66c8b746,
    },
    FunctionGUID {
        guid: 3efd6327-6038-53ed-904c-a800bee86d2e,
    },
    FunctionGUID {
        guid: 40358a63-7e16-52da-9270-434c7d444b10,
    },
    FunctionGUID {
        guid: 40358a63-7e16-52da-9270-434c7d444b10,
    },
    FunctionGUID {
        guid: 40358a63-7e16-52da-9270-434c7d444b10,
    },
    FunctionGUID {
        guid: 410b1cab-55ff-5931-a894-54842e56bee2,
    },
    FunctionGUID {
        guid: 41fd6d24-0f8b-5848-9873-65b98bd370ab,
    },
    FunctionGUID {
        guid: 4345609a-e619-5f33-84ca-7cc1ccdfa2d0,
    },
    FunctionGUID {
        guid: 43a8c54b-dd4f-5334-ae00-218cb758ad4c,
    },
    FunctionGUID {
        guid: 45770948-1c56-5b24-bd38-f5a559ae8d22,
    },
    FunctionGUID {
        guid: 45770948-1c56-5b24-bd38-f5a559ae8d22,
    },
    FunctionGUID {
        guid: 479b6feb-0dff-5c33-8f30-fd0c563b9abc,
    },
    FunctionGUID {
        guid: 4b60a419-4ad1-5808-b41f-d341868c9234,
    },
    FunctionGUID {
        guid: 4ba8a311-60f3-57d4-a17a-c0cc93f5c2a4,
    },
    FunctionGUID {
        guid: 4f0bdeb9-691c-52f8-8541-6c834f3fbe64,
    },
    FunctionGUID {
        guid: 4f645f3c-3e89-5b5d-a185-9f1a25e94ed2,
    },
    FunctionGUID {
        guid: 52b2a610-5304-5e4f-8b79-9b975e9efb86,
    },
    FunctionGUID {
        guid: 53b484f3-751a-505d-beae-acc9d69c261d,
    },
    FunctionGUID {
        guid: 53b484f3-751a-505d-beae-acc9d69c261d,
    },
    FunctionGUID {
        guid: 56c82b1f-380c-52fe-ae70-a39962e5ffdc,
    },
    FunctionGUID {
        guid: 5779439c-a22f-5d33-8868-4e1832ad49e9,
    },
    FunctionGUID {
        guid: 5b55a8c8-1d3a-5825-a0d9-917f434e07af,
    },
    FunctionGUID {
        guid: 5b7f41e9-de1e-558c-ac7d-888acca3a76c,
    },
    FunctionGUID {
        guid: 65be15d9-ef02-58a4-af63-b40ddb3e56b7,
    },
    FunctionGUID {
        guid: 669a32f2-f1e9-5b38-98aa-9eb8431ebc4e,
    },
    FunctionGUID {
        guid: 6abe3fc2-6d29-5fde-a31e-bb8249db6bf9,
    },
    FunctionGUID {
        guid: 6c01cd47-89a2-59f3-ac6a-cd7806fb8f79,
    },
    FunctionGUID {
        guid: 72452dc5-da9c-54fc-82de-032aab1780a4,
    },
    FunctionGUID {
        guid: 7995ca73-04fe-5b30-9e88-1e69197007c1,
    },
    FunctionGUID {
        guid: 7995ca73-04fe-5b30-9e88-1e69197007c1,
    },
    FunctionGUID {
        guid: 7c43c697-22ef-57a9-a43a-2214e627749d,
    },
    FunctionGUID {
        guid: 7cbd1cef-0134-5a29-a8c6-82697f5717fe,
    },
    FunctionGUID {
        guid: 7d4d1429-5fef-5c8b-9074-93e9d97ab57e,
    },
    FunctionGUID {
        guid: 82e98c80-859c-5ab3-83ab-b597fc520a3f,
    },
    FunctionGUID {
        guid: 84429ba2-9f72-585f-8cda-31e8ef4983fd,
    },
    FunctionGUID {
        guid: 86ab6ab4-45e5-5c08-9a47-05344ea503ab,
    },
    FunctionGUID {
        guid: 86ab6ab4-45e5-5c08-9a47-05344ea503ab,
    },
    FunctionGUID {
        guid: 87371a48-38c9-5e0e-81db-5e2d6da49202,
    },
    FunctionGUID {
        guid: 879b1d59-7301-5cb7-911e-db05292bee8f,
    },
    FunctionGUID {
        guid: 88d9b26b-3884-58d6-b164-435d8d888e5c,
    },
    FunctionGUID {
        guid: 89f0e45b-c1f6-5aeb-94a8-26e72bb66e6c,
    },
    FunctionGUID {
        guid: 8a2b7bda-5fdb-5ba3-888b-20d5b8d7b2cd,
    },
    FunctionGUID {
        guid: 8c736b66-b63f-5000-a060-2e60a1b8952a,
    },
    FunctionGUID {
        guid: 8c736b66-b63f-5000-a060-2e60a1b8952a,
    },
    FunctionGUID {
        guid: 8c736b66-b63f-5000-a060-2e60a1b8952a,
    },
    FunctionGUID {
        guid: 8dca0c9d-bdb7-56a1-9832-431aec16a68f,
    },
    FunctionGUID {
        guid: 9a9e3287-e407-5526-885b-1b5ee78a5300,
    },
    FunctionGUID {
        guid: 9e4c3806-b291-569c-a687-0bcaa48e248b,
    },
    FunctionGUID {
        guid: 9ef21651-3eaf-5940-8e72-104bda834e85,
    },
    FunctionGUID {
        guid: a0c07bfa-a027-5c19-9da4-d2d35d7beb40,
    },
    FunctionGUID {
        guid: a0c07bfa-a027-5c19-9da4-d2d35d7beb40,
    },
    FunctionGUID {
        guid: a23505e4-e7cd-5543-897c-46b97c1eaef3,
    },
    FunctionGUID {
        guid: a482559e-6a69-505e-b692-147deeced692,
    },
    FunctionGUID {
        guid: a482559e-6a69-505e-b692-147deeced692,
    },
    FunctionGUID {
        guid: a5b947d4-a7e0-5072-86b3-240474c16595,
    },
    FunctionGUID {
        guid: a603d813-2b33-567d-80c8-06e31562a400,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: adadfe23-cb3f-5c9f-ae8d-1dd2fd922367,
    },
    FunctionGUID {
        guid: b098f8a1-ecdd-5938-8da3-bb37935c6b7e,
    },
    FunctionGUID {
        guid: b098f8a1-ecdd-5938-8da3-bb37935c6b7e,
    },
    FunctionGUID {
        guid: b0aca296-adc3-5b46-a6b1-e82ef2100ead,
    },
    FunctionGUID {
        guid: b9111cf4-dcff-52a5-8755-2dedea6d5c72,
    },
    FunctionGUID {
        guid: bbb9b018-a0cd-5f98-b941-367f8707434d,
    },
    FunctionGUID {
        guid: bbb9b018-a0cd-5f98-b941-367f8707434d,
    },
    FunctionGUID {
        guid: bfaf2f0a-d92d-5ef7-879d-86eb0ca01257,
    },
    FunctionGUID {
        guid: bfaf2f0a-d92d-5ef7-879d-86eb0ca01257,
    },
    FunctionGUID {
        guid: c3842e85-b9f2-54d5-b353-e9bb2a0b1203,
    },
    FunctionGUID {
        guid: c3842e85-b9f2-54d5-b353-e9bb2a0b1203,
    },
    FunctionGUID {
        guid: c44bdef1-d78a-57ec-bff5-4e7d32ff4337,
    },
    FunctionGUID {
        guid: c71f4559-f0af-54dc-a224-e1f341311988,
    },
    FunctionGUID {
        guid: c71f4559-f0af-54dc-a224-e1f341311988,
    },
    FunctionGUID {
        guid: c74ed160-24fe-50d7-ba1c-9751c4599420,
    },
    FunctionGUID {
        guid: ccbd9943-6b24-59d8-ae4d-3ee868a2729c,
    },
    FunctionGUID {
        guid: ccbd9943-6b24-59d8-ae4d-3ee868a2729c,
    },
    FunctionGUID {
        guid: ce28e080-0609-5bcd-abe1-2a6e57c91f83,
    },
    FunctionGUID {
        guid: ce28e080-0609-5bcd-abe1-2a6e57c91f83,
    },
    FunctionGUID {
        guid: ce28e080-0609-5bcd-abe1-2a6e57c91f83,
    },
    FunctionGUID {
        guid: cf4808fe-2fea-517d-8f85-f88ddc96cc78,
    },
    FunctionGUID {
        guid: cf4808fe-2fea-517d-8f85-f88ddc96cc78,
    },
    FunctionGUID {
        guid: cfa4374b-0fc7-575d-9ba1-4a37d2de0406,
    },
    FunctionGUID {
        guid: d72f40b9-0557-5695-aa4d-468515ec4b8e,
    },
    FunctionGUID {
        guid: d8ba616a-1747-587c-9265-cd4e26ae35c8,
    },
    FunctionGUID {
        guid: db957206-4cca-5631-893d-1d3dc500c804,
    },
    FunctionGUID {
        guid: dd480d6b-04f7-5fb2-a7b2-236e1d78b2b0,
    },
    FunctionGUID {
        guid: e1b1e1df-fdb9-59d6-b966-5b7f926b5649,
    },
    FunctionGUID {
        guid: e480dd82-0b7d-5c75-ac67-ba8c0bb755cb,
    },
    FunctionGUID {
        guid: e4ba95bb-0221-5dbd-9884-13a39c4f9d55,
    },
    FunctionGUID {
        guid: e8290b3d-cf56-5c1b-8315-abce0f135559,
    },
    FunctionGUID {
        guid: e8290b3d-cf56-5c1b-8315-abce0f135559,
    },
    FunctionGUID {
        guid: ecca9a9d-1722-5344-b68f-54f4527c0d78,
    },
    FunctionGUID {
        guid: ed23805d-3729-590d-83a3-3b0803a56870,
    },
    FunctionGUID {
        guid: eefbc8a9-f4f1-5725-9a4f-9dcd04ea3478,
    },
    FunctionGUID {
        guid: f37337a6-4cf0-55b4-a0da-25a15e3d9bcb,
    },
    FunctionGUID {
        guid: f37cf37b-b19d-5575-b2c7-a6604d3faa69,
    },
    FunctionGUID {
        guid: f3d57592-4cb1-5356-9591-1f9327c58565,
    },
    FunctionGUID {
        guid: fb20fa7c-c796-58e4-a2f9-7a128baaf5b1,
    },
]
